### ISSUE TITLE
Map initialization fails on mobile devices 

### ISSUE THREAD

**Issue by ahokinson** [mapbox/mapbox-unity-sdk#480]
Using Unity 2017.3.0p1 and Mapbox 1.3.0 (from master), my maps will not initialize on mobile devices. They have been thoroughly tested in editor and on WebGL, but for different reasons (at least in terms of debug messages) on iOS and Android, they do not initialize.

Here is a recording of my app running on my Pixel 2: https://streamable.com/xzusi

The class I am using:

```
using System.Linq;
using Mapbox.Map;
using Mapbox.Unity.Map;
using Mapbox.Unity.Utilities;
using Mapbox.Utils;

namespace Akamai.Map
{
    public class GlobeMap : AbstractMap
    {
        protected override void Start()
        {            
            base.Start();
            var localScale = Root.localScale;
            Initialize(CenterLatitudeLongitude, (int)Zoom);
            Root.localScale = localScale;
            FindObjectsOfType<SphereCapper>().ToList().ForEach(sc => sc.GenerateCapMesh());
        }

        public override void Initialize(Vector2d latLon, int zoom)
        {   
            _worldHeightFixed = false;
            _centerLatitudeLongitude = latLon;

            var referenceTileRect = Conversions.TileBounds(TileCover.CoordinateToTileId(_centerLatitudeLongitude, zoom));
            _centerMercator = referenceTileRect.Center;

            _worldRelativeScale = (float)(_unityTileSize / referenceTileRect.Size.x);
            _mapVisualizer.Initialize(this, _fileSource);
            _tileProvider.Initialize(this);

            SendInitialized();
        }
    }
}
```

The instance of it in the inspector:
<img width="398" alt="screen shot 2017-12-30 at 7 23 14 pm" src="https://user-images.githubusercontent.com/1762048/34458191-efc9a5a6-ed96-11e7-9148-f46d80744b94.png">

Any help is much appreciated. Is there anything I should try debugging from within the Mapbox SDK to find the culprit?

**Comment by ahokinson** [comment]
Further investigation reveals that the tile provider fails during `OnInitialized()`. I will continue to debug this today and post my results here.

**Comment by ahokinson** [comment]
Because I can only debug on my device, it took quite some time to get to this point.

The path that I took to find the problem is as follows (and I learned a lot about the SDK in the process):
`GlobeMap.Initialize`
`AbstractTileProvider.Initialize`
`AbstractTileProvider.OnInitialized`
`AbstractMap.TileProvider_OnTileAdded`
`AbstractMapVisualizer.LoadTile`
`AbstractTileFactory.LoadTile`
`MapImageFactor.LoadTile`
`Tile.Initialize`
`MapboxAccess.Request`
`CachingWebFileSource.Request`
`ICache.Get`
`MbTilesCache.Get`

And finally in 
`MbTiles.Get`, I found the culprit to be LINQ expressions.

The code as it is in the SDK:
```
public byte[] GetTile(CanonicalTileId tileId)
{	
	tiles tile = _sqlite
		.Table<tiles>()
		.Where(t => t.zoom_level == tileId.Z && t.tile_column == tileId.X && t.tile_row == tileId.Y)
		.FirstOrDefault();

	if (null == tile)
	{
		return null;
	}

	return tile.tile_data;
}
```

The code which resolved my null reference exception:
```
public byte[] GetTile(CanonicalTileId tileId)
{	
	tiles tile = _sqlite
		.Table<tiles>()
		.FirstOrDefault(t => t.zoom_level == tileId.Z && t.tile_column == tileId.X && t.tile_row == tileId.Y);

	if (null == tile)
	{
		return null;
	}

	return tile.tile_data;
}
```

Though I have solved that null reference exception, I continue to have issues with the map initialization on mobile. I will continue to work on this and update the issue tomorrow.

**Comment by eusebiu** [comment]
Did you include this file into the project? https://github.com/mapbox/mapbox-unity-sdk/blob/master/utils/link.xml

**Comment by ahokinson** [comment]
It is included in the SDK at Mapbox/Core/Plugins/link.xml, so yes.

**Comment by atripathi-mb** [comment]
@ahokinson I would suggest using the `OnInitialized`  event on the map to trigger any processes that need the map to be loaded. The issue you are seeing could be caused because the map is being used before being fully initialized. 

The LINQ related change should not be necessary

**Comment by ahokinson** [comment]
Why would it work in editor, desktop, and WebGL? How is what I'm doing any different than in `AbstractMap.cs`?

```
protected virtual void Start()
{
	if (_initializeOnStart)
	{
		Initialize(Conversions.StringToLatLon(_latitudeLongitudeString), AbsoluteZoom);
	}
	_initialZoom = AbsoluteZoom;
}
```

**Comment by atripathi-mb** [comment]
@ahokinson 
>Why would it work in editor, desktop, and WebGL? 

Unity does not guarantee the order in which objects are created, my initial thought was that you might be using the map before it is fully ready for use, looking at your call stack and code snippet that doesn't seem to be the case. 

Here are a few thing I would suggest to try out in order for us better understand the issue. 
1. Try one of the SDK examples on the mobile device to check if the tiles load. This would ensure that the network permissions and access tokens are set correctly. 
2. Do you have the `Tile Error Handler` script attached to your map? With v1.3.0 the tile errors are reported if this script is attached to the map, this might help narrow the specific errors if any. 

cc @BergWerkGIS - Do you have any insights on what the issue might be?

**Comment by BergWerkGIS** [comment]
@ahokinson 
hm, that Linq error is weird. 
Your change shouldn't make any difference, except in the case the `tiles` table doesn't exist at all, which shouldn't happen - at all.
Maybe the database file is corrupt.

Could you try the following on Android:
* go into Android Settings -> Apps -> <YOUR_APP> -> Storage:
  * tap `CLEAR DATA`
  * tap `CLEAR CACHE`
* delete your app
* re-deploy

Does it work?

My second hunch is also about permissions but it's strange that this happens on Android and on iOS.

Any chance we could get our hands on a small test case that demos the problem?

**Comment by ahokinson** [comment]
I was able to resolve the issue by altering the two LINQ statements inside of `MbTiles.cs`.

You can see here that it is working: https://streamable.com/sy8gw

```
/// <summary>
/// Returns the tile data, otherwise null
/// </summary>
/// <param name="tileId">Canonical tile id to identify the tile</param>
/// <returns>tile data as byte[], if tile is not cached returns null</returns>
public byte[] GetTile(CanonicalTileId tileId)
{	
	tiles tile = _sqlite
		.Table<tiles>()
		.FirstOrDefault(t => t.zoom_level == tileId.Z && t.tile_column == tileId.X && t.tile_row == tileId.Y);

	if (null == tile)
	{
		return null;
	}

	return tile.tile_data;
}


/// <summary>
/// Check if tile exists
/// </summary>
/// <param name="tileId">Canonical tile id</param>
/// <returns>True if tile exists</returns>
public bool TileExists(CanonicalTileId tileId)
{
	return null != _sqlite
		.Table<tiles>()
		.FirstOrDefault(t => t.zoom_level == tileId.Z && t.tile_column == tileId.X && t.tile_row == tileId.Y);
}
```

I would also say that the two LINQ statements, while similar, are not equivalent.
My best guess is that the SQLite library that you're using doesn't implement the `Table` enumerable in an expected way resulting in some bizarre null in my particular setup.

I would gladly provide my project (in some capacity) for further testing, though it would take me some time to strip it down to a minimal example size.

Since this fix resolves the issue, I would be willing to submit a pull request for the changes.

**Comment by ahokinson** [comment]
I'm actually going to keep this open and ask another question, since the above fix only appears to fix Android.

On iOS, I get all the tiles to load, but I get an exception:
```
InvalidTokenException: StatusNotYetSet
```

Any ideas? The globe loads, but remains untextured.

**Comment by BergWerkGIS** [comment]
> Since this fix resolves the issue, I would be willing to submit a pull request for the changes.
> I'm actually going to keep this open and ask another question, since the above fix only appears to fix Android.

Thanks for offering a PR with your changes, but that's not a real fix as it is just hiding the actual problem: table `tiles` does not exist in the cache database - and that should never happen.

There seems to be problem with `sqlite` not being able to properly create a cache database.

Do you get any other errors despite the Linq one?
Related to caching or `sqlite`?

----

Did you try my suggestion [from above](https://github.com/mapbox/mapbox-unity-sdk/issues/480#issuecomment-355139719) on Android?

Did that change anything?

----

> On iOS, I get all the tiles to load, but I get an exception: `InvalidTokenException: StatusNotYetSet`

This indicates a network problem, meaning your access token could not be verified.

----

> I would gladly provide my project (in some capacity) for further testing, though it would take me some time to strip it down to a minimal example size.

A test project would really help as I'm tapping in the dark not being able to reproduce your issues.
And we are dealing with at least two of them: cache failure and token verification.

You could send a download link to `<MY-FIRST-NAME> AT mapbox DOT com>` to share privately and confidentially.

**Comment by ahokinson** [comment]
I cannot try your solution on Android because the app is uninstalled and reinstalled between all testing I've done.

My debugging workflow was:
```
adb uninstall com.akamai.gio
adb install gio.apk
```

That should remove all data related to the app and install fresh, so there would be no data or cache to clear.

I get no other errors on Android.

I will try more iOS debugging related to the network this morning.

And I'd be willing to share my entire project if you're willing to try it out, let me just run it by my supervisor and get it to you later today.

**Comment by ahokinson** [comment]
Okay, so the iOS issue was due to a super out of date test phone we have in the office. We updated it and now it works the same as Android. Thanks for the hint about network issues!

I also spoke with my supervisor and despite my opinion, he said that I couldn't send the whole application. So instead, I'm going to strip everything out of it, but leave the map as I have it. So it'll basically be the Mapbox SDK and my implementation of the Abstract Map class.
